bitvm/src/signatures/public.rs

```diff
@@ -0,0 +1,838 @@
+        let secret_key = WOTS::generate_secret_key();
+        test_vectors.push(create_test_vector::<WOTS>(&message, secret_key));
+
+        // Sign all-zeroes message where signature has trailing zero byte
+        'grind_trailing_zero: loop {
+            let secret_key = WOTS::generate_secret_key();
+            let witness = WOTS::sign_to_raw_witness(&secret_key, &message);
+            for entry in &witness {
+                if entry.last().map(|&byte| byte == 0).unwrap_or(false) {
+                    test_vectors.push(create_test_vector::<WOTS>(&message, secret_key));
+                    break 'grind_trailing_zero;
+                }
+            }
+        }
+
+        // Sign all-zeroes message where signature has leading zero byte
+        'grind_leading_zero: loop {
+            let secret_key = WOTS::generate_secret_key();
+            let witness = WOTS::sign_to_raw_witness(&secret_key, &message);
+            for entry in &witness {
+                if entry.first().map(|&byte| byte == 0).unwrap_or(false) {
+                    test_vectors.push(create_test_vector::<WOTS>(&message, secret_key));
+                    break 'grind_leading_zero;
+                }
+            }
+        }
+
+        // Sign all-ones message
+        let message = WOTS::Message::try_from(vec![0xff; MSG_BYTE_LEN]).unwrap();
+        let secret_key = WOTS::generate_secret_key();
+        test_vectors.push(create_test_vector::<WOTS>(&message, secret_key));
+
+        // Sign random message
+        let message: Vec<u8> = (0..MSG_BYTE_LEN).map(|_| rand::random()).collect();
+        let message = WOTS::Message::try_from(message).unwrap();
+        let secret_key = WOTS::generate_secret_key();
+        test_vectors.push(create_test_vector::<WOTS>(&message, secret_key));
+
+        // Sign half-zeroes, half-ones message
+        let message: Vec<u8> = (0..MSG_BYTE_LEN / 2)
+            .map(|_| 0x00u8)
+            .chain((0..MSG_BYTE_LEN / 2).map(|_| 0xffu8))
+            .collect();
+        let message = WOTS::Message::try_from(message).unwrap();
+        let secret_key = WOTS::generate_secret_key();
+        test_vectors.push(create_test_vector::<WOTS>(&message, secret_key));
+
+        // Sign third-zeroes, third-ones, third-random message
+        let message: Vec<u8> = (0..MSG_BYTE_LEN / 3)
+            .map(|_| 0x00u8)
+            .chain((0..MSG_BYTE_LEN / 3).map(|_| 0xffu8))
+            .chain((0..MSG_BYTE_LEN - MSG_BYTE_LEN / 3 * 2).map(|_| rand::random()))
+            .collect();
+        let message = WOTS::Message::try_from(message).unwrap();
+        let secret_key = WOTS::generate_secret_key();
+        test_vectors.push(create_test_vector::<WOTS>(&message, secret_key));
+
+        test_vectors
+    }
+
+    // Uncomment and run this code to generate fresh test vectors:
+    // cargo test generate_and_save_test_vectors
+    /*
+    #[test]
+    fn generate_and_save_test_vectors() -> io::Result<()> {
```

Comment:
Can this be flagged as ignored instead?

Reply:
There are already ignored tests that we expect to run via `cargo test -- --ignored`. If generating test vectors is an ignored test, then it will overwrite the existing test vectors. Someone with little knowledge of git may accidentally commit 6000 lines of changed JSON.

Reply:
Got it. Would be nice to have this as a separate executable, so it would be subject to checks in the event of future changes. Is that reasonable to do?

Reply:
I put the test vector generator next to the test vector verifier so it's easy to modify later. Adding a separate executable increases complexity, which seems unnecessary to me. I don't expect frequent updates to the test vectors.